fe2-android/app/src/androidTest/java/com/github/dedis/popstellar/utility/security/KeyManagerTest.java

```diff
@@ -0,0 +1,136 @@
+
+    assertEquals(mainKey, mainKeyPair.getPublicKey());
+
+    // We cannot extract the public key from the keyset handle
+    // But we can make sure both signatures are equals and the key manager keypair can verify it
+    PublicKeySign signer = androidKeysetManager.getKeysetHandle().getPrimitive(PublicKeySign.class);
+    // Generate any data, here a message id
+    Base64URLData data = Base64DataUtils.generateMessageID();
+
+    Signature signature = mainKeyPair.sign(data);
+    assertArrayEquals(signer.sign(data.getData()), signature.getData());
+    assertTrue(mainKey.verify(signature, data));
+  }
+
+  @Test
+  public void popTokenRetrievingWorks() throws KeyException {
+    PoPToken token = Base64DataUtils.generatePoPToken();
+    when(wallet.recoverKey(any(), any(), any())).thenReturn(token);
+
+    // create LAO and RollCalls
+    Lao lao = new Lao("lao", Base64DataUtils.generatePublicKey(), 54213424);
+    RollCall rollCall1 = new RollCall(lao.getId(), 5421364, "rollcall1");
+    RollCall rollCall2 = new RollCall(lao.getId(), 5421363, "rollcall2");
+
+    lao.updateRollCall(rollCall1.getId(), rollCall1);
+    lao.updateRollCall(rollCall2.getId(), rollCall2);
+
+    KeyManager manager = new KeyManager(androidKeysetManager, wallet);
+    assertEquals(token, manager.getValidPoPToken(lao));
+    assertEquals(token, manager.getValidPoPToken(lao, rollCall1));
+
+    // make sure that rollcall1 was taken and not rollcall2 as the oldest is rollcall 1
+    verify(wallet).recoverKey(eq(lao.getId()), eq(rollCall1.getId()), any());
+  }
+
+  @Test
+  public void popTokenRetrievingFailsWhenWalletFails() throws KeyException {
+    PoPToken token = Base64DataUtils.generatePoPToken();
+
+    // create LAO and RollCall
+    Lao lao = new Lao("lao", Base64DataUtils.generatePublicKey(), 54213424);
+    RollCall rollCall = new RollCall(lao.getId(), 5421364, "rollcall");
+    lao.updateRollCall(rollCall.getId(), rollCall);
+
+    KeyManager manager = new KeyManager(androidKeysetManager, wallet);
+
+    // Test with every possible errors
+    when(wallet.recoverKey(any(), any(), any()))
+        .thenThrow(new KeyGenerationException(new GeneralSecurityException()));
+    assertThrows(KeyGenerationException.class, () -> manager.getValidPoPToken(lao));
+    verify(wallet, times(1)).recoverKey(eq(lao.getId()), eq(rollCall.getId()), any());
+    reset(wallet);
+
+    when(wallet.recoverKey(any(), any(), any())).thenThrow(new UninitializedWalletException());
+    assertThrows(UninitializedWalletException.class, () -> manager.getValidPoPToken(lao));
+    verify(wallet, times(1)).recoverKey(eq(lao.getId()), eq(rollCall.getId()), any());
+    reset(wallet);
+
+    when(wallet.recoverKey(any(), any(), any())).thenThrow(new InvalidPoPTokenException(token));
+    assertThrows(InvalidPoPTokenException.class, () -> manager.getValidPoPToken(lao));
+    verify(wallet, times(1)).recoverKey(eq(lao.getId()), eq(rollCall.getId()), any());
```

Comment:
in general, it's better for checks like this to have their own separate test cases